File: 34-Search-for-a-Range/solution.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> results = {-1, -1};
        int size = nums.size();
        int left = 0;
        int right = size-1;
        if(left==right && results[0]==target) {
            results[0]=results[1]=0;
            return results;
        }
        int mid = right;
        while(left < right) {
            if(nums[mid]==target){
                left = right = mid;
                break;
            } else if(nums[mid] < target) {
                left = mid+1;
            } else {
                right = mid-1;
            }
            mid=left+(right-left)/2;
        }
        if(nums[left]==target) {
            while(left>0 && nums[left-1]==target) left--;
            while(right<size-1 && nums[right+1]==target) right++;
            results[0] = left;
            results[1] = right;
        }
        return results;
    }
};

File: 34-Search-for-a-Range/solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] ans = new int[]{-1, -1};
        if (nums == null || nums.length == 0) {
            return ans;
        }
        int lo = 0, hi = nums.length - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        if (nums[lo] != target) {
            return ans;
        }
        ans[0] = lo;
        hi = nums.length - 1;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2 + 1;
            if (nums[mid] > target) {
                hi = mid - 1;
            } else {
                lo = mid;
            }
        }
        ans[1] = lo;
        return ans;
    }
}